### ISSUE TITLE
HLSL SV_GroupId argument with uint/uint2 type generates invalid SPIR-V.

### ISSUE THREAD

**Issue by danginsburg** [KhronosGroup/glslang#1371]
The below HLSL shader (which is valid in D3D11) demonstrates the problem:

```
RWTexture2D < float4 > OutputTexture;

[ numthreads ( 8 , 8 , 1 ) ] 
void MainCs ( uint2 vGroupId : SV_GroupID ) 
{ 
    OutputTexture[ vGroupId . xy ] = float4(1.0, 0.0, 0.0, 1.0); 
} 
```

The output of glslangValidator.exe -D -e MainCs -V -s comp -o groupid_bug.spv groupid_bug.hlsl produces:

```
; SPIR-V
; Version: 1.0
; Generator: Khronos Glslang Reference Front End; 6
; Bound: 35
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %MainCs "MainCs" %vGroupId
               OpExecutionMode %MainCs LocalSize 8 8 1
               OpSource HLSL 500
               OpName %MainCs "MainCs"
               OpName %OutputTexture "OutputTexture"
               OpName %vGroupId "vGroupId"
               OpDecorate %OutputTexture DescriptorSet 0
               OpDecorate %vGroupId BuiltIn WorkgroupId
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
       %uint = OpTypeInt 32 0
     %v2uint = OpTypeVector %uint 2
      %float = OpTypeFloat 32
    %v4float = OpTypeVector %float 4
    %float_1 = OpConstant %float 1
    %float_0 = OpConstant %float 0
         %19 = OpConstantComposite %v4float %float_1 %float_0 %float_0 %float_1
         %20 = OpTypeImage %float 2D 0 0 0 2 Rgba32f
%_ptr_UniformConstant_20 = OpTypePointer UniformConstant %20
%OutputTexture = OpVariable %_ptr_UniformConstant_20 UniformConstant
%_ptr_Input_v2uint = OpTypePointer Input %v2uint
   %vGroupId = OpVariable %_ptr_Input_v2uint Input
     %MainCs = OpFunction %void None %3
          %5 = OpLabel
         %29 = OpLoad %v2uint %vGroupId
         %34 = OpLoad %20 %OutputTexture
               OpImageWrite %34 %29 %19
               OpReturn
               OpFunctionEnd
```

Note that  %vGroupId = OpVariable %_ptr_Input_v2uint Input is invalid, the spec says:

```
“The variable decorated with code:WorkgroupId must: be declared as a
three-component vector of 32-bit integers.”
```

I believe glslang needs to declare this as a uint3 even though the shader has it declared as a uint2.  spirv-val does not catch this, I will file a bug there too referencing this one.  NVIDIA 397.31 driver crashes compiling the shader though so a fairly serious bug.

**Comment by antiagainst** [comment]
FWIW, we have handled similar cases in DXC:

https://github.com/Microsoft/DirectXShaderCompiler/blob/857724e895280aa022f575f0d7bcebc8e42cabeb/tools/clang/lib/SPIRV/DeclResultIdMapper.cpp#L1282-L1284

And it's not only `SV_GroupID`, we also need similar handling for `SV_DispatchThreadID` and `SV_GroupThreadID`.

**Comment by LoopDawg** [comment]
Tentative fix in #1379, but only for the SV_GroupId example from @danginsburg.  I will look at the others from @antiagainst soon and then remove the WIP.

Sorry it took a while.

**Comment by LoopDawg** [comment]
Added SV_DispatchThreadID and SV_GroupThreadID, removed WIP.
